promremote/client.go

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package promremote
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"time"
+
+	"github.com/m3db/m3/src/query/ts"
+
+	"github.com/golang/protobuf/proto"
+	"github.com/golang/snappy"
+	"github.com/prometheus/prometheus/prompb"
+)
+
+const (
+	// DefaultRemoteWrite is the default Prom remote write endpoint in m3coordinator.
+	DefaultRemoteWrite = "http://localhost:7201/api/v1/prom/remote/write"
+
+	defaulHTTPClientTimeout = time.Second * 30
+)
+
+// Tag are the metric tags.
+type Tag struct {
+	Name  string `json:"name"`
+	Value string `json:"value"`
+}
+
+// Timeseries are made of tags and a datapoint.
+type Timeseries struct {
```

Comment:
does this need struct tags as well? if not, does `Tag` actually need them?

Reply:
Yeah, perhaps remove the `json:"name"` etc fields since they don't look to be used.

Also nit: Can you name this `TimeSeries` to match the Prometheus proto name for time series?

Reply:
Yeah they aren't very consistent with their naming convention: e.g.
```
type WriteRequest struct {
	Timeseries []*TimeSeries `protobuf:"bytes,1,rep,name=timeseries" json:"timeseries,omitempty"`
}
```

Reply:
in the absence of anything else, let's stick with the industry - `TimeSeries`, "time series", etc